src/util/createServerRootMixin.js

```diff
@@ -0,0 +1,255 @@
+import Vue from 'vue';
+import _renderToString from 'vue-server-renderer/basic';
+import instantsearch from 'instantsearch.js/es';
+import algoliaHelper from 'algoliasearch-helper';
+const { SearchResults, SearchParameters } = algoliaHelper;
+import { warn } from './warn';
+
+function walkIndex(indexWidget, visit) {
+  visit(indexWidget);
+
+  return indexWidget.getWidgets().forEach(widget => {
+    if (widget.$$type !== 'ais.index') return;
+    visit(widget);
+    walkIndex(widget, visit);
+  });
+}
+
+function renderToString(app) {
+  return new Promise((resolve, reject) =>
+    _renderToString(app, (err, res) => {
+      if (err) reject(err);
+      resolve(res);
+    })
+  );
+}
+
+function searchOnlyWithDerivedHelpers(helper) {
+  return new Promise((resolve, reject) => {
+    helper.searchOnlyWithDerivedHelpers();
+
+    // we assume all derived helpers resolve at least in the same tick
+    helper.derivedHelpers[0].on('result', e => {
+      resolve(e);
+    });
+
+    helper.derivedHelpers.forEach(derivedHelper =>
+      derivedHelper.on('error', e => {
+        reject(e);
+      })
+    );
+  });
+}
+
+function augmentInstantSearch(instantSearchOptions, searchClient, indexName) {
+  /* eslint-disable no-param-reassign */
+
+  const helper = algoliaHelper(searchClient, indexName);
```

Comment:
Can you explain why we need to create a new helper instance (and cannot rely on `search.mainIndex.getHelper()`)?

Reply:
helper is still `null` before init is called, right?

Reply:
Yeah, that's right!